README.md
#angular-basicauth

[![Build Status](https://travis-ci.org/iandotkelly/angular-basicauth.svg?branch=master)](https://travis-ci.org/iandotkelly/angular-basicauth)

[![Coverage Status](https://img.shields.io/coveralls/iandotkelly/angular-basicauth.svg)](https://coveralls.io/r/iandotkelly/angular-basicauth?branch=master)

[![Dependency Status](https://gemnasium.com/iandotkelly/angular-basicauth.svg)](https://gemnasium.com/iandotkelly/angular-basicauth)

## Purpose
Simple AngularJs service for managing login and logout, and protection
of server API endpoints using basic authentication. It provides
the following features:

- Service for handling login and logout operations
- Validates login against a configurable server endpoint
defaulting to /api/authentication
- Promise API for success/error on login
- Broadcasts 'login', 'logout' and 'authentication-failed' events on $rootScope
to allow all parts of the application to respond to auth events
- Once logged in provides an $http interceptor to add Authorization http header
to all requests
- Stores username and authentication credentials in local-storage or
cookies as a fallback
- Automatically logs user out on failed authentication
- Automatic logout after 3 hours (configurable)

Basic authentication sends usernames and passwords unencrypted in each request, so
should only be used over a HTTPS protected request.

**Please note this module stores unencypted username & password in local-storage or cookies, so is
susceptible to attack from code from the same domain**

**As with all security and authentication related software, I strongly encourage the user of this
software to thoroughly review it and understand how it work and assure themselves that it is
suitable before using it in their own project**



## Installation

You can clone this repository and include angular-basicauth.js or
angular-basicauth.min.js as a script in your application, plus its dependencies.

Alternatively you can use [bower.js](http://bower.io/) to install, which will install the
module and all its dependencies.

```sh
bower install angular-basicauth
```
## Use

### Include scripts

You should include the following scripts that comprise the module
and its bower installed dependencies.

 * angular-basicauth.js
 * bower_components/angular-base64/angular-base64.js
 * bower_components/angular-local-storage/dist/angular-local-storage.js

... or the minified equivalents.

### Add dependency in your Angular.Js app

Include the script in your angular.js web application:

```javascript
var app = angular.module('myApp', ['angularBasicAuth']);
```

### Configure Authentication URL

The authentication URL is any URL that will process a simple
GET request and return 200 if the user/password combination
encoded in the Basic Authentication header authenticates.
Any other response will be assumed to be a failure in login.

By default this URL is /api/authenticate

You can modify the URL by modifying the defaults, by injecting the
authDefaults values object, for example:

```javascript
var controller = app.controller('MyController',
	[
		'authDefaults',
		function(authDefaults) {
			// modify the auth service URL
			authDefaults.authenticateUrl = '/my/other/api.aspx';
		}
	]);
```

### Login / Logout

You can use the service to manage login & logout, by injecting
authService into your controller, for example this controller has
two methods on its scope for handling login and logout actions (say
from buttons in your UI) and listens to the broadcast events for
login/logout events that happen elsewhere.


```javascript
var controller = app.controller('MyController',
	[
		'authService', // the authentication service
		'$rootScope',  // (optional) if you want to receive auth events
		'$scope',
		function(authService, $rootScope, $scope) {

			// define the endpoints that will be authenticated
			authService.addEndpoint(); // the current hostname
			authService.addEndpoint('https://some.other.hostname.com');

			// listen for login events
			$rootScope.$on('login', function() {
				$scope.loggedInUsername = authService.username();
			});

			// listen for logout events
			$rootScope.$on('logout', function() {
				$scope.loggedInUsername = null;
			});

			// method to log-in
			$scope.onLoginButton = function () {
				// pass input username and password to
				// the service for authentication
				authService
				.login($scope.username, $scope.password)
				.success(function() {
					// handle login success
				})
				.error(function() {
					// handle login error
				});
			};

			// method to log out
			$scope.onLogoutButton = function () {
				// simply call the logout button
				authService.logout();
			};
		}
	]);
```

### HTTP Intercept

When logged in, requests generated with the $http service
have the 'Authorization' header appended to them so that
every request is authenticated.  

Please note that this means you MUST TRUST the endpoint of
these requests.

Only hostnames that have been added with a call to addEndpoint()
will have an authentication header, e.g.:

```javascript
// authenticate all requests to the current hostname
authService.addEndpoint();

// authenticate another hostname
authService.addEndpoint('https://some.other.domain.com');
```

### Custom Headers

You can add custom headers to all requests through adding them
to the headers object, for example:

```javascript
authService.headers['custom-header'] = 'custom-value';
```

## License

The MIT License (MIT)

Copyright (c) 2014 Ian Kelly

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


history.md

### 0.3.0 - October 5, 2014

- Added capability to include custom headers in requests
- Updated dependencies

### 0.2.1 - July 20, 2014

- Rebuilt minified version, otherwise identical to 0.2.0

### 0.2.0 - July 20, 2014

- Add list of protected endpoints, to prevent authorization header
being added to unnecessary requests, potentially leaking
credentials
- Change how $rootscope events from $emit to $broadcast
- Updated dependencies.  Tested against AngularJs 1.2.20

### 0.1.1 - July 1, 2014

- Add minified build
- Add to docs

### 0.1.0 - June 30, 2014

- First working version


test/test.js

'use strict';

describe('basic auth service', function () {

	describe('MODULE_VERSION', function() {

		var MODULE_VERSION;

		beforeEach(module('angularBasicAuth'));

		beforeEach(inject(function($injector) {
			MODULE_VERSION = $injector.get('MODULE_VERSION');
		}));

		it('should be a string', function () {
			expect(typeof MODULE_VERSION).toBe('string');
		});

	});

	describe('authDefaults', function() {

		var authDefaults;

		beforeEach(module('angularBasicAuth'));

		beforeEach(inject(function ($injector) {
			authDefaults = $injector.get('authDefaults');
		}));

		it('should be an object', function () {
			expect(typeof authDefaults).toBe('object');
		});

		it('should have an authenticateUrl property', function() {
			expect(authDefaults.authenticateUrl).toBeDefined();
			expect(typeof authDefaults.authenticateUrl).toBe('string');
		});

		it('should have an sessonMinutes property', function() {
			expect(authDefaults.sessionMinutes).toBeDefined();
			expect(typeof authDefaults.sessionMinutes).toBe('number');
		});

	});

	describe('authService', function() {

		describe('uninitialized', function() {

			var authService;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				authService = $injector.get('authService');
				authService.addEndpoint();
			}));

			it('should be an onbject', function () {
				expect(typeof authService).toBe('object');
			});

			it('shoud have a defined api', function () {
				expect(typeof authService.addEndpoint).toBe('function');
				expect(typeof authService.isEndpoint).toBe('function');
				expect(typeof authService.login).toBe('function');
				expect(typeof authService.logout).toBe('function');
				expect(typeof authService.username).toBe('function');
				expect(typeof authService.getAuth).toBe('function');
				expect(typeof authService.activity).toBe('function');
				expect(typeof authService.handleAuthFailure).toBe('function');
			});

		});

		describe('before credentials', function() {

			var authService;
			var $httpBackend;
			var $http;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				$httpBackend = $injector.get('$httpBackend');
				$httpBackend.resetExpectations();
				$http = $injector.get('$http');
				authService = $injector.get('authService');
				authService.addEndpoint();
			}));

			afterEach(function() {
				$httpBackend.verifyNoOutstandingExpectation ();
				$httpBackend.verifyNoOutstandingRequest ();
				$httpBackend.resetExpectations();
			});

			it('#getAuth should return null', function() {
				expect(authService.getAuth()).toBeNull();
			});

			it('#username should return null', function() {
				expect(authService.username()).toBeNull();
			});

			it('an http request should not have an Authorization header', function() {

				$httpBackend.expectGET('/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBeUndefined();
					return [200, ''];
				});

				$http.get('/fred');
				$httpBackend.flush();
			});
		});

		describe('logging in with incorrect credentials', function() {

			var authService;
			var server;
			var $rootScope;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				authService = $injector.get('authService');
				authService.addEndpoint();
				$rootScope = $injector.get('$rootScope');

				// cannot use $httpBackend as the service bypasses $http
				// for the authenticate calls to avoid a circular reference
				// so use the sinon fake server API
				server = sinon.fakeServer.create();
			}));

			afterEach(function() {
				server.restore();
			});

			it('should invoke a request to /api/authenticate with a header', function(done) {

				server.respondWith(function(request) {
					expect(request.url).toBe('/api/authenticate');
					expect(request.requestHeaders.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					request.respond(401, {}, '');
					done();
				});

				authService.login('ian@me', 'fred');
				server.respond();
			});

			it('should call the error function of the promise', function(done) {

				server.respondWith(function(request) {
					request.respond(401, {}, '');
				});

				authService.login('ian@me', 'fred').success(function() {
					// this should not be called
					expect(true).toBe(false);
					done();
				}).error(function() {
					// hurrah
					expect(authService.username()).toBeNull();
					expect(authService.getAuth()).toBeNull();
					done();
				});

				server.respond();
				$rootScope.$digest();
			});

			it('should emit the correct events from $rootScope', function(done) {

				server.respondWith(function(request) {
					request.respond(401, {}, '');
				});

				authService.login('ian@me', 'fred');

				$rootScope.$on('login', function() {
					// this should not be called
					expect(true).toBe(false);
					done();
				});

				$rootScope.$on('authentication-failed', function() {
					done();
				});

				server.respond();
				$rootScope.$digest();
			});
		});

		describe('changing the url', function() {

			var authService;
			var authDefaults;
			var server;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				authService = $injector.get('authService');
				authService.addEndpoint();
				authDefaults = $injector.get('authDefaults');

				// cannot use $httpBackend as the service bypasses $http
				// for the authenticate calls to avoid a circular reference
				// so use the sinon fake server API
				server = sinon.fakeServer.create();
			}));

			afterEach(function() {
				server.restore();
			});

			it('should invoke a request to /a/different/url.aspx with a header', function(done) {

				// change the URL
				authDefaults.authenticateUrl = '/a/different/url.aspx';

				server.respondWith(function(request) {
					expect(request.url).toBe('/a/different/url.aspx');
					expect(request.requestHeaders.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					request.respond(401, {}, '');
					done();
				});

				authService.login('ian@me', 'fred');
				server.respond();
			});
		});

		describe('logging in with correct credentials', function() {

			var authService;
			var server;
			var $rootScope;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				authService = $injector.get('authService');
				authService.addEndpoint();
				$rootScope = $injector.get('$rootScope');

				// cannot use $httpBackend as the service bypasses $http
				// for the authenticate calls to avoid a circular reference
				// so use the sinon fake server API
				server = sinon.fakeServer.create();
			}));

			afterEach(function() {
				server.restore();
			});

			it('should invoke a request to /api/authenticate with a header', function(done) {

				server.respondWith(function(request) {
					expect(request.requestHeaders.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					request.respond(200, {}, '');
					done();
				});

				authService.login('ian@me', 'fred');
				server.respond();
			});

			it('should call the success function of the promise and record user', function(done) {

				server.respondWith(function(request) {
					request.respond(200, {}, '');
				});

				authService.login('ian@me', 'fred').success(function() {
					// hurrah
					expect(authService.username()).toBe('ian@me');
					expect(authService.getAuth()).toBe('Basic aWFuQG1lOmZyZWQ=');
					done();
				}).error(function() {
					// this should not be called
					expect(true).toBe(false);
					done();
				});

				server.respond();
				$rootScope.$digest();
			});

			it('should emit the correct events from $rootScope', function(done) {

				server.respondWith(function(request) {
					request.respond(200, {}, '');
				});

				authService.login('ian@me', 'fred');

				$rootScope.$on('login', function() {
					done();
				});

				$rootScope.$on('authentication-failed', function() {
					// this should not be called
					expect(true).toBe(false);
					done();
				});

				server.respond();
				$rootScope.$digest();
			});
		});

		describe('adding headers to the authService', function() {

			var authService;
			var server;
			var $rootScope;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				authService = $injector.get('authService');
				authService.addEndpoint();
				authService.headers['test-header'] = 'this is a test';
				$rootScope = $injector.get('$rootScope');

				// cannot use $httpBackend as the service bypasses $http
				// for the authenticate calls to avoid a circular reference
				// so use the sinon fake server API
				server = sinon.fakeServer.create();
			}));

			afterEach(function() {
				server.restore();
			});

			it('should invoke a request to /api/authenticate with a custom header', function(done) {

				server.respondWith(function(request) {
					expect(request.requestHeaders.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					expect(request.requestHeaders['test-header']).toBe('this is a test');
					request.respond(200, {}, '');
					done();
				});

				authService.login('ian@me', 'fred');
				server.respond();
			});
		});

		describe('after valid credentials', function() {

			var authService;
			var $httpBackend;
			var $http;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				$httpBackend = $injector.get('$httpBackend');
				$httpBackend.resetExpectations();
				$http = $injector.get('$http');
				authService = $injector.get('authService');
				authService.addEndpoint();
				authService.headers['custom-header'] = 'custom value';
			}));

			afterEach(function() {
				$httpBackend.verifyNoOutstandingExpectation ();
				$httpBackend.verifyNoOutstandingRequest ();
				$httpBackend.resetExpectations();
			});

			it('#getAuth should return the auth string', function() {
				expect(authService.getAuth()).toBe('Basic aWFuQG1lOmZyZWQ=');
			});

			it('#username should return the username', function() {
				expect(authService.username()).toBe('ian@me');
			});

			it('an http request should have an Authorization header', function() {

				$httpBackend.expectGET('/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					return [200, ''];
				});

				$http.get('/fred');
				$httpBackend.flush();
			});

			it('custom headers should also be included in the request', function() {

				$httpBackend.expectGET('/fred').respond(function(method, url, data, headers) {
					expect(headers['custom-header']).toBe('custom value');
					return [200, ''];
				});

				$http.get('/fred');
				$httpBackend.flush();
			});

			it('an http request to a different hostname should not have an Authorization header', function() {

				$httpBackend.expectGET('http://fred.com/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBe(undefined);
					return [200, ''];
				});

				$http.get('http://fred.com/fred');
				$httpBackend.flush();
			});

			it('until a new endpoint is added when it should have an Authorization header', function() {

				authService.addEndpoint('http://fred.com/');

				$httpBackend.expectGET('http://fred.com/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					return [200, ''];
				});

				$http.get('http://fred.com/fred');
				$httpBackend.flush();
			});

			it('adding an endpoint more than once does not cause a problem', function() {

				authService.addEndpoint('http://fred.com/');
				authService.addEndpoint('http://fred.com/');

				$httpBackend.expectGET('http://fred.com/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					return [200, ''];
				});

				$http.get('http://fred.com/fred');
				$httpBackend.flush();
			});

		});



		describe('calling #logout()', function() {

			var authService;
			var $httpBackend;
			var $http;
			var $rootScope;
			var server;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				$httpBackend = $injector.get('$httpBackend');
				$httpBackend.resetExpectations();
				$http = $injector.get('$http');
				$rootScope = $injector.get('$rootScope');
				authService = $injector.get('authService');
				authService.addEndpoint();

				// cannot use $httpBackend as the service bypasses $http
				// for the authenticate calls to avoid a circular reference
				// so use the sinon fake server API
				server = sinon.fakeServer.create();
			}));

			beforeEach(function(done) {
				// perform a login
				server.respondWith(function(request) {
					request.respond(200, {}, '');
				});

				authService.login('ian@me', 'fred')
				.success(function() {
					var interval = setInterval(function() {
						// wait for the digest to complete before calling done();
						if (!$rootScope.$$phase) {
							clearInterval(interval);
							done();
						}
					}, 5);
				})
				.error(function() {
					throw 'Failed to login';
				});

				server.respond();
				$rootScope.$digest();
			});

			afterEach(function() {
				$httpBackend.verifyNoOutstandingExpectation ();
				$httpBackend.verifyNoOutstandingRequest ();
				$httpBackend.resetExpectations();
				server.restore();
			});

			it('should remove the credentials', function() {
				expect(authService.username()).toBe('ian@me');
				expect(authService.getAuth()).toBe('Basic aWFuQG1lOmZyZWQ=');
				authService.logout();
				expect(authService.username()).toBeNull();
				expect(authService.getAuth()).toBeNull();
			});


			it('should broadcast a logout event on $rootScope', function(done) {

				$rootScope.$on('logout', function() {
					done();
				});

				authService.logout();
			});

			it('an http request should not have an Authorization header', function() {

				authService.logout();

				$httpBackend.expectGET('/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBeUndefined();
					return [200, ''];
				});

				$http.get('/fred');
				$httpBackend.flush();
			});
		});


		describe('after an authentication failure on a request', function() {

			var authService;
			var $httpBackend;
			var $http;
			var $rootScope;
			var server;

			beforeEach(module('angularBasicAuth'));

			beforeEach(inject(function ($injector) {
				$httpBackend = $injector.get('$httpBackend');
				$httpBackend.resetExpectations();
				$http = $injector.get('$http');
				$rootScope = $injector.get('$rootScope');
				authService = $injector.get('authService');
				authService.addEndpoint();

				// cannot use $httpBackend as the service bypasses $http
				// for the authenticate calls to avoid a circular reference
				// so use the sinon fake server API
				server = sinon.fakeServer.create();
			}));

			beforeEach(function(done) {
				// perform a login
				server.respondWith(function(request) {
					request.respond(200, {}, '');
				});

				authService.login('ian@me', 'fred')
				.success(function() {
					var interval = setInterval(function() {
						// wait for the digest to complete before calling done();
						if (!$rootScope.$$phase) {
							clearInterval(interval);
							done();
						}
					}, 5);
				})
				.error(function() {
					throw 'Failed to login';
				});

				server.respond();
				$rootScope.$digest();
			});

			afterEach(function() {
				$httpBackend.verifyNoOutstandingExpectation ();
				$httpBackend.verifyNoOutstandingRequest ();
				$httpBackend.resetExpectations();
				server.restore();
			});

			it('it should invoke a logout event', function(done) {

				$rootScope.$on('logout', function() {
					// i have no idea why this calling done() can only happen
					// when the difest is over
					var interval = setInterval(function() {
						// wait for the digest to complete before calling done();
						if (!$rootScope.$$phase) {
							clearInterval(interval);
							done();
						}
					}, 5);
				});

				$httpBackend.expectGET('/fred').respond(function(method, url, data, headers) {
					expect(headers.Authorization).toBe('Basic aWFuQG1lOmZyZWQ=');
					return [401, ''];
				});

				$http.get('/fred');
				$httpBackend.flush();
			});
		});
	});
});


test/karma.conf.js
// Karma configuration
// Generated on Sun Jun 29 2014 19:32:02 GMT-0400 (EDT)
'use strict';

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '..',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
        'bower_components/angular/angular.js',
        'bower_components/angular-mocks/angular-mocks.js',
        'bower_components/angular-local-storage/dist/angular-local-storage.js',
        'bower_components/angular-base64/angular-base64.js',
        'bower_components/sinonjs/sinon.js',
        'angular-basicauth.js',
        'test/**/*.js',
    ],

    // list of files to exclude
    exclude: [

    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'angular-basicauth.js': 'coverage'
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['spec', 'coverage'],

    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true
  });
};


Gruntfile.js
/**
 * Gruntfile for Angular-Basic-Auth
 */

'use strict';

module.exports = function (grunt) {

	require('load-grunt-tasks')(grunt);

	// Project configuration.
	grunt.initConfig({

		jshint: {
			// general jshint settings
			all: {
				options: {
					jshintrc: '.jshintrc'
				},
				src: [
					'Gruntfile.js',
					'angular-basicauth.js'
				]
			}
		},

		uglify: {
			dist: {
				files: {
					'angular-basicauth.min.js': 'angular-basicauth.js'
				}
			}
		},

		karma: {
			chrome: {
				configFile: 'test/karma.conf.js'
			},
			travis: {
				configFile: 'test/karma-phantom.conf.js'
			}
		}

	});

	grunt.registerTask('test', ['jshint', 'karma:chrome']);
	grunt.registerTask('lint', ['jshint']);
	grunt.registerTask('dist', ['jshint', 'uglify']);
	grunt.registerTask('travis', ['jshint', 'karma:travis']);
};


angular-basicauth.js
/**
 * Basic Authentication Module for AngularJs
 *
 * Copyright (C) Ian Kelly
 * License: MIT
 */

'use strict';

(function() {

	var angularBasicAuth = angular.module(
		'angularBasicAuth',
		[
			'LocalStorageModule',
			'base64'
		]);

	// Constants
	var MS_PER_MINUTE = 1000 * 60;

	/**
	* Difference between two dates in hours
	*
	* @param {Date} a First date
	* @param {Date} b Second date
	*/
	function dateDiffInMinutes(earlier, later) {
		return (later.getTime() - earlier.getTime()) / MS_PER_MINUTE;
	}

	/**
	 * Constants
	 */
	angularBasicAuth.constant('MODULE_VERSION', '0.0.0');

	/**
	 * Defaults
	 */
	angularBasicAuth.value('authDefaults', {
		authenticateUrl: '/api/authenticate',
		sessionMinutes: 180
	});

	/**
	* HTTP Interceptor to add Authorization headers to all requests if the
	* auth local storage property has been set
	*/
	angularBasicAuth.factory('authInterceptor',
		[
			'$q',
			'authService',
			function ($q, authService) {

				var isEndpoint = authService.isEndpoint;

				var authInterceptor = {

					/**
					* Request Interceptor
					*
					* Adds an authorization header, if and only if there
					* is one found in the local storage
					*/
					request: function(config) {
						if (isEndpoint(config.url)) {
							var auth = authService.getAuth();
							var headers = authService.headers;
							if (auth) {
								config.headers.Authorization = auth;
								for (var key in headers) {
									if (headers.hasOwnProperty(key)) {
										config.headers[key] = headers[key];
									}
								}
							}
						}
						return config;
					},

					/**
					* Error Interceptor
					*
					* Used to trap an authentication error
					*/
					responseError: function(rejection) {
						// this is a failure to authenticate
						// so inform the authenticaton service
						if (rejection.status === 401) {
							authService.handleAuthFailure();
						}

						return $q.reject(rejection);
					}
				};

				return authInterceptor;
			}
		]);

	/**
	* Service to handle user authentication
	*
	* This is a singleton service, so we will instantiate this once in an iife
	*/
	(function() {
		angularBasicAuth.factory('authService',
			[
				'$log',
				'localStorageService',
				'$base64',
				'$q',
				'$rootScope',
				'$interval',
				'authDefaults',
				function ($log, localStorage, $base64, $q, $rootScope, $interval, authDefaults) {

					$log.debug('authService constructed');

					var LS_USERNAME = 'username';
					var LS_AUTHENTICATION = 'auth';
					var LS_LASTACTIVITY = 'last-activity';

					// array to store protected endpoints
					var endpoints = [];
					// use an anchor element to parse urls
					var urlParser = document.createElement('a');

					/**
					 * Whether a hostname is in the endpoint collection
					 * @param {String} hostname  The hostname of a URL, e.g. www.google.com
					 */
					function inEndpoints(hostname) {
						for (var index = 0, len = endpoints.length; index < len; index++) {
							if (hostname === endpoints[index]) {
								return true;
							}
						}
						return false;
					}

					/**
					* Logout and wipe the local storage
					*/
					function logout() {
						$log.debug('logout event');
						localStorage.remove(LS_AUTHENTICATION);
						localStorage.remove(LS_LASTACTIVITY);
						localStorage.remove(LS_USERNAME);
						$rootScope.$emit('logout');
					}

					/**
					* Set the user credentials
					*
					* @param {String} username Username
					* @param {String} password Unencryped password
					*/
					function setCredentials(username, password) {
						$log.debug('Setting credentials for user: ' + username);

						localStorage.set(LS_USERNAME, username);

						// set the value of the auth-header
						localStorage.set(LS_AUTHENTICATION,
							'Basic ' + $base64.encode(username + ':' + password));

						recordActivity();
					}

					/**
					* Record account activity
					*/
					function recordActivity() {
						localStorage.set(LS_LASTACTIVITY, (new Date()).toString());
					}

					/**
					* Is the login current
					*/
					function isCurrent() {
						// retrieve the last activity of the account
						var lastActivity = localStorage.get(LS_LASTACTIVITY) || '';

						if (lastActivity === '') {
							return false;
						}

						return (dateDiffInMinutes(new Date(lastActivity), new Date()) < authDefaults.sessionMinutes);
					}

					/**
					* Confirm the account is current
					*/
					function confirmCurrent() {
						if (!isCurrent()) {
							$log.debug('Authentication credentials missing or out of date');
							logout();
						}
					}

					/**
					* Returns the current authentication header
					*/
					function getAuth() {
						return localStorage.get(LS_AUTHENTICATION);
					}

					// we should work out whether we are current or not
					confirmCurrent();

					// set an inteval to do this regularly
					$interval(confirmCurrent, 60000);

					return {
						/**
						* Get any current authentication header
						*/
						getAuth: getAuth,

						/**
						* Handles logout
						*/
						logout: logout,

						/**
						* Record activity
						*/
						activity: recordActivity,

						/**
						 * Add an endpoint
						 *
						 * @param {String} endpoint  The URL of an endpoint, or undefined if current location
						 */
						addEndpoint: function(endpoint) {
							endpoint = endpoint || window.location;
							urlParser.href = endpoint;
							var hostname = urlParser.hostname;
							if (inEndpoints(hostname)) {
								return;
							}
							endpoints.push(hostname);
						},

						/**
						 * Additional headers to add to authenticated
						 * requests.
						 */
						headers: {},

						/**
						 * Is an URL a defined endpoint requiring authentication
						 *
						 * @param {String} url The URL of a potential endpoint
						 */
						isEndpoint: function(url) {
							urlParser.href = url;
							var hostname = urlParser.hostname;
							return inEndpoints(hostname);
						},

						/**
						* The current username
						*
						* @return {String} The current username
						*/
						username: function() {
							confirmCurrent();
							return localStorage.get(LS_USERNAME);
						},

						/**
						* Handle an authentication failure
						*/
						handleAuthFailure: function() {
							logout();
							$rootScope.$emit('authentication-failure');
						},

						/**
						* Handles login to the application
						*
						* @param  {String} username The username
						* @param  {String} password The password
						* @return {Object}          Promise for the transaction
						*/
						login: function(username, password, headers) {
							$log.debug('login event');

							// record the credentials
							setCredentials(username, password);

							var deferred = $q.defer();

							/**
							* Process the response, and if there are any issues
							* the reject the promise
							*/
							function processResponse() {
								/* jshint validthis: true */
								if (this.status !== 200) {
									deferred.reject();
								} else {
									deferred.resolve();
								}
							}

							// don't use $http to avoid circular reference, so
							// directly use XMLHttpRequest and $q
							var promise = deferred.promise;
							var request = new XMLHttpRequest();
							request.onload = processResponse;
							request.open('GET', authDefaults.authenticateUrl);
							request.setRequestHeader('Accept', 'application/json');
							request.setRequestHeader('Authorization', getAuth());
							for (var key in this.headers) {
							  if (this.headers.hasOwnProperty(key)) {
									request.setRequestHeader(key, this.headers[key]);
							  }
							}
							request.send(null);

							/**
							* Allow subscribers to the promise to add a
							* success function to the login event
							*/
							promise.success = function(fn) {
								promise.then(function() {
									fn();
								});
								return promise;
							};

							/**
							* Allow subscribers to the promise to add
							* an error function to the login event
							*/
							promise.error = function(fn) {
								promise.then(null, function() {
									fn();
								});
								return promise;
							};

							/**
							* Handle the promise being rejected or resolved
							*/
							promise.then(function () {
								// we've confirmed credentials match a user
								$log.debug('Successfully authenticated');
								$rootScope.$broadcast('login', username);
							}, function () {
								// some error in credential check
								$log.debug('Test authentication failed');
								logout();
								$rootScope.$broadcast('authentication-failed', username);
							});

							return promise;
						}
					};
				}
			]
		);
	})();

	// add the interceptor to the http service
	angularBasicAuth.config(['$httpProvider', function($httpProvider) {
		$httpProvider.interceptors.push('authInterceptor');
	}]);

})();
